File: packages/papyrus-lang/src/program/DisplayTextEmitter.ts
import { StringBuilder } from '../common/StringBuilder';
import { Flags, isLiteral, NodeKind } from '../parser/Node';
import {
    CustomEventSymbol,
    EventSymbol,
    FunctionSymbol,
    GroupSymbol,
    ImportSymbol,
    ParameterSymbol,
    PropertySymbol,
    ScriptSymbol,
    StateSymbol,
    Symbol,
    SymbolKind,
    VariableSymbol,
} from '../symbols/Symbol';
import { Program } from './Program';

export interface DisplayText {
    kind: string;
    text: string;
    documentation?: string;
}

export interface EventDisplayText {
    kind: string;
    prefix: string;
    parameters: DisplayText[];
    documentation?: string;
}

export interface FunctionDisplayText {
    kind: string;
    prefix: string;
    shortNamePrefix: string;
    parameters: DisplayText[];
    postfix: string;
    documentation?: string;
}

export class DisplayTextEmitter {
    private readonly _program: Program;

    constructor(program: Program) {
        this._program = program;
    }

    public getDisplayText(symbol: Symbol): DisplayText {
        switch (symbol.kind) {
            case SymbolKind.Function:
                const funcDisplayText = this.getDisplayTextForFunction(symbol);
                return {
                    kind: funcDisplayText.kind,
                    text: `${
                        funcDisplayText.prefix
                    }(${funcDisplayText.parameters
                        .map((p) => p.text)
                        .join(', ')})${funcDisplayText.postfix}`,
                    documentation: funcDisplayText.documentation,
                };
            case SymbolKind.Event:
                const eventDisplayText = this.getDisplayTextForEvent(symbol);
                return {
                    kind: eventDisplayText.kind,
                    text: `${
                        eventDisplayText.prefix
                    }(${eventDisplayText.parameters
                        .map((p) => p.text)
                        .join(', ')})`,
                    documentation: eventDisplayText.documentation,
                };
            case SymbolKind.Script:
                return this.getDisplayTextForScript(symbol);
            case SymbolKind.State:
                return this.getDisplayTextForState(symbol);
            case SymbolKind.Variable:
                return this.getDisplayTextForVariable(symbol);
            case SymbolKind.Property:
                return this.getDisplayTextForProperty(symbol);
            case SymbolKind.Parameter:
                return this.getDisplayTextForParameter(symbol);
            case SymbolKind.CustomEvent:
                return this.getBasicDisplayText(
                    symbol.fullyQualifiedName,
                    'custom event',
                    'CustomEvent'
                );
            case SymbolKind.Group:
                return this.getBasicDisplayText(
                    symbol.fullyQualifiedName,
                    'group',
                    'Group'
                );
            case SymbolKind.Import:
                return this.getBasicDisplayText(
                    symbol.name,
                    'import',
                    'Import'
                );
            case SymbolKind.Struct:
                return this.getBasicDisplayText(
                    symbol.fullyQualifiedName,
                    'struct',
                    'Struct'
                );
            default:
                return null;
        }
    }

    public getDisplayTextForEvent(symbol: EventSymbol): EventDisplayText {
        const prefix = new StringBuilder();

        prefix.append(`Event ${symbol.fullyQualifiedName}`);

        return {
            kind: 'event',
            prefix: prefix.toString(),
            parameters: symbol.parameters.map((p) =>
                this.getDisplayTextForParameter(p)
            ),
            documentation: symbol.documentation,
        };
    }

    public getDisplayTextForFunction(
        symbol: FunctionSymbol
    ): FunctionDisplayText {
        const prefix = new StringBuilder();
        const shortNamePrefix = new StringBuilder();

        if (symbol.returnType) {
            const returnType = this._program.getTypeForName(
                symbol.returnType.name
            );
            prefix.append(
                `${
                    returnType
                        ? returnType.fullyQualifiedName
                        : symbol.returnType.name
                } `
            );
        }

        shortNamePrefix.append(prefix.toString());

        prefix.append(`Function ${symbol.fullyQualifiedName}`);
        shortNamePrefix.append(`Function ${symbol.name}`);

        const postfix = new StringBuilder();

        if (symbol.isGlobal) {
            postfix.append(' Global');
        }

        if (symbol.isNative) {
            postfix.append(' Native');
        }

        if (symbol.userFlags.length > 0) {
            postfix.append(' ');
            postfix.append(symbol.userFlags.join(' '));
        }

        return {
            kind: 'function',
            prefix: prefix.toString(),
            shortNamePrefix: shortNamePrefix.toString(),
            parameters: symbol.parameters.map((p) =>
                this.getDisplayTextForParameter(p)
            ),
            postfix: postfix.toString(),
            documentation: symbol.documentation,
        };
    }

    private getDisplayTextForVariable(symbol: VariableSymbol) {
        const sb = new StringBuilder();

        const valueType = this._program.getTypeForName(symbol.valueType.name);
        sb.append(
            `${
                valueType ? valueType.fullyQualifiedName : symbol.valueType.name
            } `
        );

        sb.append(
            symbol.declaration.node.kind === NodeKind.DeclareStatement
                ? symbol.name
                : symbol.fullyQualifiedName
        );

        if (
            symbol.declaration.node.initialValue &&
            isLiteral(symbol.declaration.node.initialValue)
        ) {
            sb.append(` = ${symbol.declaration.node.initialValue.value}`);
        }

        if (symbol.isConst) {
            sb.append(' IsConst');
        }

        return {
            kind: 'variable',
            text: sb.toString(),
        };
    }

    private getDisplayTextForParameter(symbol: ParameterSymbol) {
        const sb = new StringBuilder();

        const valueType = this._program.getTypeForName(symbol.valueType.name);
        sb.append(
            `${
                valueType ? valueType.fullyQualifiedName : symbol.valueType.name
            } `
        );

        sb.append(symbol.name);

        if (symbol.declaration.node.defaultValue) {
            sb.append(` = ${symbol.declaration.node.defaultValue.value}`);
        }

        return {
            kind: 'parameter',
            text: sb.toString(),
        };
    }

    private getDisplayTextForProperty(symbol: PropertySymbol) {
        const sb = new StringBuilder();

        const valueType = this._program.getTypeForName(symbol.valueType.name);
        sb.append(
            `${
                valueType ? valueType.fullyQualifiedName : symbol.valueType.name
            } `
        );

        sb.append(`Property ${symbol.fullyQualifiedName}`);

        if (symbol.declaration.node.initialValue) {
            sb.append(` = ${symbol.declaration.node.initialValue.value}`);
        }

        if (symbol.isAutoReadOnly) {
            sb.append(' AutoReadOnly');
        } else if (symbol.isAuto) {
            sb.append(' Auto');
        }

        if (symbol.isConst) {
            sb.append(' Const');
        }

        return {
            kind: 'property',
            text: sb.toString(),
            documentation: symbol.documentation,
        };
    }

    private getDisplayTextForScript(symbol: ScriptSymbol): DisplayText {
        const sb = new StringBuilder();
        sb.append(`Scriptname ${symbol.name}`);

        if (symbol.extendedScript) {
            const scriptType = this._program.getTypeForName(
                symbol.extendedScript.name
            );

            sb.append(
                ` extends ${
                    scriptType ? scriptType.name : symbol.extendedScript.name
                }`
            );
        }

        if (symbol.isConst) {
            sb.append(' Const');
        }

        if (symbol.isNative) {
            sb.append(' Native');
        }

        if (symbol.userFlags.length > 0) {
            sb.append(' ');
            sb.append(symbol.userFlags.join(' '));
        }

        return {
            kind: 'script',
            text: sb.toString(),
            documentation: symbol.documentation,
        };
    }

    private getBasicDisplayText(name: string, kind: string, keyword: string) {
        const sb = new StringBuilder();
        sb.append(`${keyword} ${name}`);

        return {
            kind,
            text: sb.toString(),
        };
    }

    private getDisplayTextForState(symbol: StateSymbol) {
        const sb = new StringBuilder();
        if (symbol.isAuto) {
            sb.append('Auto ');
        }

        sb.append(`State ${symbol.fullyQualifiedName}`);

        return {
            kind: 'state',
            text: sb.toString(),
        };
    }
}

File: src/DarkId.Papyrus.LanguageService/Program/Symbols/DisplayTextEmitter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DarkId.Papyrus.Common;
using DarkId.Papyrus.LanguageService.Program.Syntax;

namespace DarkId.Papyrus.LanguageService.Program.Symbols
{
    public class DisplayText
    {
        public string Kind { get; set; }
        public string Text { get; set; }
        public string Documentation { get; set; }
    }

    public class FunctionDisplayText
    {
        public string Kind { get; set; }
        public string Prefix { get; set; }
        public string ShortNamePrefix { get; set; }
        public List<DisplayText> Parameters { get; set; }
        public string Postfix { get; set; }
        public string Documentation { get; set; }
    }

    public class DisplayTextEmitter
    {
        public string GetEventHandlerSignatureText(PapyrusSymbol symbol)
        {
            if (symbol is EventSymbol eventSymbol)
            {
                var displayText = GetDisplayTextForEvent(eventSymbol);
                return $"{eventSymbol.Name}({symbol.Parent.Name} akSender, {displayText.Parameters.Select(t => t.Text).Join(", ")})";
            }

#if FALLOUT4
            if (symbol is CustomEventSymbol customEventSymbol)
            {
                return $"{customEventSymbol.Name}({symbol.Parent.Name} akSender, Var[] akArgs)";
            }
#endif
            return null;
        }

        public DisplayText GetDisplayText(PapyrusSymbol symbol)
        {
            if (symbol is AliasedSymbol asAliased)
            {
                return GetDisplayText(asAliased.Aliased);
            }

            // TODO: User flags

            switch (symbol.Kind)
            {
                case SymbolKinds.CustomEvent:
                    return GetBasicDisplayText(symbol.Name, "custom event", "CustomEvent");
                case SymbolKinds.Event:
                    var eventDisplayText = GetDisplayTextForEvent((EventSymbol)symbol);
                    return new DisplayText()
                    {
                        Kind = eventDisplayText.Kind,
                        Text = $"{eventDisplayText.Prefix}({eventDisplayText.Parameters.Select(t => t.Text).Join(", ")}){eventDisplayText.Postfix}",
                        Documentation = eventDisplayText.Documentation
                    };
                case SymbolKinds.Function:
                    var funcDisplayText = GetDisplayTextForFunction((FunctionSymbol)symbol);
                    return new DisplayText()
                    {
                        Kind = funcDisplayText.Kind,
                        Text = $"{funcDisplayText.Prefix}({funcDisplayText.Parameters.Select(t => t.Text).Join(", ")}){funcDisplayText.Postfix}",
                        Documentation = funcDisplayText.Documentation
                    };
                case SymbolKinds.Group:
                    return GetBasicDisplayText(symbol.Name, "group", "Group");
                case SymbolKinds.Import:
                    return GetBasicDisplayText(symbol.Name, "import", "Import");
                case SymbolKinds.Property:
                    return GetDisplayTextForProperty((PropertySymbol)symbol);
                case SymbolKinds.State:
                    return GetDisplayTextForState((StateSymbol)symbol);
                case SymbolKinds.Struct:
                    return GetBasicDisplayText(
                        symbol.GetPapyrusType()?.Name.FullyQualifiedDisplayName
                        ?? symbol.Name, "struct", "Struct");
                case SymbolKinds.Variable:
                    return GetDisplayTextForVariable((VariableSymbol)symbol);
                case SymbolKinds.Script:
                    return GetDisplayTextForScript((ScriptSymbol)symbol);
            }

            return null;
        }

        private DisplayText GetBasicDisplayText(string name, string kind, string keyword)
        {
            var sb = new StringBuilder();
            sb.Append($"{keyword} {name}");

            return new DisplayText()
            {
                Kind = kind,
                Text = sb.ToString()
            };
        }

        private DisplayText GetDisplayTextForState(StateSymbol symbol)
        {
            var sb = new StringBuilder();

            if ((symbol.Flags & LanguageFlags.Auto) != 0)
            {
                sb.Append("Auto ");
            }

            sb.Append(symbol.Name);

            return new DisplayText()
            {
                Kind = "state",
                Text = sb.ToString(),
                Documentation = symbol.GetDocumentationMarkdown()
            };
        }

        private DisplayText GetDisplayTextForScript(ScriptSymbol symbol)
        {
            var sb = new StringBuilder();

            var scriptType = symbol.GetPapyrusType();
            sb.Append($"Scriptname {scriptType?.Name.FullyQualifiedDisplayName ?? symbol.Name}");

            if (symbol.ExtendedScript != null)
            {
                var extendedType = symbol.ExtendedScript.GetPapyrusType();
                sb.Append(" extends " + extendedType?.Name.FullyQualifiedDisplayName ?? symbol.ExtendedScript.Name);
            }

            if (symbol.Flags != LanguageFlags.None)
            {
                sb.Append(" ");
                sb.Append(symbol.Flags.ToString().Replace(",", string.Empty));
            }

            return new DisplayText()
            {
                Kind = "script",
                Text = sb.ToString(),
                Documentation = symbol.GetDocumentationMarkdown()
            };
        }

        private DisplayText GetDisplayTextForVariable(VariableSymbol symbol)
        {
            var sb = new StringBuilder();

            var valueType = symbol.GetPapyrusType();
            sb.Append(valueType != null ? valueType.Name.FullyQualifiedDisplayName.ToString() : symbol.Definition.TypeIdentifier.Text);
            sb.Append(" ");

            if (symbol.Parent is TypeSymbol typeSymbol)
            {
                sb.Append(typeSymbol.GetPapyrusType()?.Name.ShortName ?? typeSymbol.Name);
                sb.Append(".");
            }

            sb.Append(symbol.Name);

            // TODO: Default value

            return new DisplayText()
            {
                Kind = ((SyntaxNode)symbol.Definition).Parent is FunctionHeaderNode ? "parameter" : "variable",
                Text = sb.ToString(),
                Documentation = symbol.GetDocumentationMarkdown()
            };
        }

        private DisplayText GetDisplayTextForProperty(PropertySymbol symbol)
        {
            var sb = new StringBuilder();

            var valueType = symbol.GetPapyrusType();
            sb.Append(valueType != null ? valueType.Name.FullyQualifiedDisplayName : symbol.Definition.Header.TypeIdentifier.Text);
            sb.Append(" Property ");

            sb.Append($"{symbol.Script.Id.ShortName}.{symbol.Name}");

            if (symbol.Flags != LanguageFlags.None)
            {
                sb.Append(" ");
                sb.Append(symbol.Flags.ToString().Replace(",", string.Empty));
            }

            return new DisplayText()
            {
                Kind = "property",
                Text = sb.ToString(),
                Documentation = symbol.GetDocumentationMarkdown()
            };
        }

        public FunctionDisplayText GetDisplayTextForEvent(EventSymbol symbol)
        {
            return new FunctionDisplayText()
            {
                Kind = "event",
                Prefix = $"Event {symbol.Script.Id}.{symbol.Name}",
                ShortNamePrefix = $"Event {symbol.Script.Id.ScriptName}.{symbol.Name}",
                Parameters = symbol.Definition.Header.Parameters.Select(p =>
                    GetDisplayTextForVariable((VariableSymbol)p.Symbol)).ToList(),
                Documentation = symbol.GetDocumentationMarkdown()
            };
        }

        public FunctionDisplayText GetDisplayTextForFunction(FunctionSymbol symbol)
        {
            var prefix = new StringBuilder();
            var shortNamePrefix = new StringBuilder();

            var typeIdentifier = symbol.Definition.Header.TypeIdentifier;
            if (typeIdentifier != null)
            {
                var returnType = symbol.GetPapyrusType();
                prefix.Append(returnType?.Name.FullyQualifiedDisplayName ?? typeIdentifier.Text + (typeIdentifier.IsArray ? "[]" : string.Empty));
                prefix.Append(" ");
            }

            shortNamePrefix.Append(prefix.ToString());

            prefix.Append($"Function {symbol.Script.Id}.{symbol.Name}");
            shortNamePrefix.Append($"Function {symbol.Script.Id.ScriptName}.{symbol.Name}");

            var postfix = new StringBuilder();

            if (symbol.Flags != LanguageFlags.None)
            {
                postfix.Append(" ");
                postfix.Append(symbol.Flags.ToString().Replace(",", string.Empty));
            }

            return new FunctionDisplayText()
            {
                Kind = "function",
                Prefix = prefix.ToString(),
                ShortNamePrefix = shortNamePrefix.ToString(),
                Parameters = symbol.Definition.Header.Parameters.Select(p =>
                    GetDisplayTextForVariable((VariableSymbol)p.Symbol)).ToList(),
                Postfix = postfix.ToString(),
                Documentation = symbol.GetDocumentationMarkdown()
            };
        }
    }
}
